zain/quiz/README.md

```diff
@@ -0,0 +1,589 @@
+1. What's the output?
+
+function sayHi() {
+  console.log(name);
+  console.log(age);
+  var name = 'Lydia';
+  let age = 21;
+}
+
+sayHi();
+
+A: Lydia and undefined
+
+B: Lydia and ReferenceError
+
+C: ReferenceError and 21
+
+D: ~~undefined and ReferenceError~~
+
+In the sayHi function, both name and age are accessed before they are declared. name is declared with var, so it's hoisted and initialized with undefined, allowing it to be accessed before its actual declaration. However, age is declared with let, and accessing it before its declaration results in a ReferenceError due to the temporal dead zone.
+
+
+2. What's the output?
+
+for (var i = 0; i < 3; i++) {
+  setTimeout(() => console.log(i), 1);
+}
+
+for (let i = 0; i < 3; i++) {
+  setTimeout(() => console.log(i), 1);
+}
+
+A: 0 1 2 and 0 1 2
+
+B: 0 1 2 and 3 3 3
+
+C: ~~3 3 3 and 0 1 2~~
+
+In the first loop, var creates a single variable i with function scope. The asynchronous setTimeout captures the final value of i, which is 3, resulting in 3 3 3 being printed.
```

Comment:
Variables declared with var are hoisted, and in each iteration, there's a setTimeout function, meaning these console.logs are executed in the task queue. Once the call stack is executed and cleared, these console.logs are pushed into the call stack by the event loop, but by this time, the for loop has already finished and incremented the value of i to three.